### ISSUE TITLE
Implementing Typegoose possible

### ISSUE THREAD

**Issue by smolinari** [doug-martin/nestjs-query#645]
Hi,

I was just wondering if it would make sense to take the time to also do a Typegoose implementation/ package. I believe Typegoose (and the NestJs-Typegoose module) is a bit more flexible than Nest's own Mongoose module in that more can be done in/ with the entities. This is also an opinion based only on very high-level research of issues in both packages. 

I've not dug in at all into the nestjs-query-mongoose package and thought I'd ask before I do. Well, not too much. I did see there was Typegoose mentioned in a test case. 😁 

At any rate, would a Typegoose implementation make sense?

Scott

**Comment by smolinari** [comment]
I'm also trying to run the unit tests and running into a slew of strange fails. Seems like the tests with any kind of number is expecting negative values and getting positive ones. I can't find anything describing that kind of issue with Jest using google either. 

🤷 

Scott

**Comment by smolinari** [comment]
There are so many tests failing it must be something general to either my setup or Jest's setup. 

I'm running Win 10 Pro. Node 12.19.0. 

This is the final result. 

![image](https://user-images.githubusercontent.com/3314957/97795907-2cb24d00-1c0c-11eb-857d-202ab3385759.png)

All the tests failing seem to be `toEqual` with either text or objects being compared. Most of them are the SDL to schema created, where it looks like there is no issue with the text. The text looks correct between the two results. I read somewhere about `toEqual` failing due to hidden characters. I'll continue to look for the issue. 

Very strange. :)

Scott

**Comment by smolinari** [comment]
Ok. As it seems, it's a Windows/ Git issue. 

This is the text with `JSON.stringify`

```
console.log
      "type Query {\n  updateTest(input: UpdateOne!): Int!\n}\n\ninput UpdateOne {\n  \"\"\"The id of the record to update\"\"\"\n  id: ID!\n\n  \"\"\"The update to apply.\"\"\"\n  update: FakeUpdateOneType!\n}\n\ninput FakeUpdateOneType {\n  name: String!\n}\n"

      at Object.<anonymous>.exports.expectSDL (packages/query-graphql/__tests__/__fixtures__/index.ts:22:11)

    console.log
      "type Query {\r\n  updateTest(input: UpdateOne!): Int!\r\n}\r\n\r\ninput UpdateOne {\r\n  \"\"\"The id of the record to update\"\"\"\r\n  id: ID!\r\n\r\n  \"\"\"The update to apply.\"\"\"\r\n  update: FakeUpdateOneType!\r\n}\r\n\r\ninput FakeUpdateOneType {\r\n  name: String!\r\n}\r\n"
```

If Git for Windows isn't set up correctly, it will add `/r/n` as line breaks when cloning a repo. If the project was developed in anything other than Windows, then text comparisons with `toEqual` that have line breaks in them will fail. Too bad Jest doesn't say something like "mismatched line breaks" or show the escaped characters.

After fixing the issue with Git. 

![image](https://user-images.githubusercontent.com/3314957/97796662-178dec00-1c15-11eb-94c7-00ea1252912b.png)


Scott

**Comment by smolinari** [comment]
I'll also try what you suggested @doug-martin and report back. If Typegoose does work too, I'll close this issue.

Scott

**Comment by smolinari** [comment]
Ok. Nope. Typegoose isn't 1 to 1 compatible. I'll try my luck at building out a Typegoose equivalent as an additional package. Typegoose is being well maintained and I do think it has some advantages, like just a bit more documentation and some additional  features available within the TypeScript side of it all. By that I mean, one doesn't have to go into the Mongoose schema side of things to do, for instance, discriminators. You can define discriminators within the ModelOptions decorator, which I believe nestjs-mongoose doesn't have. 

If it's ok, I'll leave this issue open, and if I can get the Typegoose package working, I'll PR it (and if you are interested). I'll make sure to add docs too. :)

By the way, your docs are still showing Typegoose example code. :)

https://kpfromer.github.io/nestjs-typegoose/docs/usage#creating-the-service

Scott

**Comment by JohnMcInall** [comment]
Hi @smolinari , 

I have a typegoose implementation of this library that I have been building over the last couple of weeks, the only thing I still need to do it write the full documentation for it but you are welcome to try it out :)

You can find a built version of it here - https://github.com/JohnMcInall/nestjs-query-typegoose

I will be making a PR in the coming days after I write up the documentation to do with it, but if you would like to take a look about beforehand please feel free!

Some screenshot to show a basic implementation for now:

Module:
![image](https://user-images.githubusercontent.com/24442639/97917382-4d81bc00-1d4c-11eb-81eb-944dede134f5.png)

DTO:
![image](https://user-images.githubusercontent.com/24442639/97917444-625e4f80-1d4c-11eb-9bea-cc84e61f8b52.png)

Entity: (Base comes from Typegoose types, its give you _id property and some other stuff)
![image](https://user-images.githubusercontent.com/24442639/97917596-99346580-1d4c-11eb-8ca8-018772e5a6e6.png)


Any questions please feel free to reach out and I will do my best to assist!

**Comment by smolinari** [comment]
Hi @JohnMcInall,

At first site, this looks great!! :D Thanks for chiming in, as I definitely don't want to do double work/ reinvent the wheel. 😁 Any chance I can get into the repo with the source code to help? I don't mind doing docs as I'm not the greatest programmer either. 😊 

@doug-martin - would be great, if you could chime in as to if you even want this as an extra package. 😃  

Scott

**Comment by JohnMcInall** [comment]
@smolinari Hey, I only need to write the e2e tests in the example stuff now, you can find the source code on the forked branch on my account [here](https://github.com/JohnMcInall/nestjs-query) :+1:

**Comment by smolinari** [comment]
@JohnMcInall - Are you in Discord? I'm available as Scott (EN-DE)#0052.

Scott

**Comment by JohnMcInall** [comment]
@smolinari I am, I have added you :+1:
